04-linear-chain.Rmd

```diff
@@ -0,0 +1,576 @@
+```{r, include = FALSE}
+ottrpal::set_knitr_image_path()
+```
+
+# Connecting multiple tasks together in a linear chain
+
+Now that you have a first task in a workflow up and running, the next step is to continue building out the workflow as described in our Workflow Plan:
+
+1.  `BwaMem` aligns the samples to the reference genome.
+2.  `MarkDuplicates` marks PCR duplicates.
+3.  `ApplyBaseRecalibrator` applies base recalibration.
+4.  `Mutect2` performs somatic mutation calling. For this current iteration, we start with `Mutect2TumorOnly` which only uses the tumor sample.
+5.  `annovar` annotates the called somatic mutations.
+
+We do this via a **linear chain**, in which we feed the output of one task to the input of the next task. Let's see how to build a linear chain in a workflow.
+
+## How to connect tasks together in a workflow
+
+We can easily connect tasks together because of the following: *The output variables of a task can be accessed at the workflow level as inputs for the subsequent task.*
+
+For instance, in this dummy `LinearChain` workflow below, after the task `stepA` is called, its task output `out` can be accessed at the workflow level as inputs for the task `stepB` in the form of `stepA.out`. This example is adapted from [OpenWDL Docs](https://docs.openwdl.org/en/latest/WDL/Linear_chaining/)' introduction to linear chain, which has more [introductory examples and variations of writing linear chains](https://docs.openwdl.org/en/latest/WDL/add_plumbing/).
+
+```         
+version 1.0
+workflow LinearChain {
+  input {
+    File firstInput
+  }
+  call stepA { 
+    input: 
+      in = firstInput
+  }
+  call stepB {
+   input:
+     in = stepA.out
+  }
+}
+
+task stepA {
+  input {
+    File in
+  }
+  command <<<
+    programA I=~{in} O=outputA.ext
+  >>>
+  output { 
+    File out = "outputA.ext"
+  }
+}
+
+task stepB {
+  input {
+    File in
+  }
+  command <<<
+    programB I=~{in} O=outputB.ext
+  >>>
+  output { 
+    File out = "outputB.ext"
+  }
+}
+```
+
+Now let's see how this applies to our workflow: we will take the output of `BwaMem` task into the input of the `MarkDuplicates` task. Recall that in the `BwaMem` task, the **output** section of the task is the following:
+
+```         
+output {
+    File analysisReadyBam = "~{base_file_name}.aligned.bam"
+    File analysisReadySorted = "~{base_file_name}.sorted_query_aligned.bam"
+  }
+```
+
+The File variables `analysisReadyBam` and `analysisReadySorted` can now be accessed anywhere in the workflow block after the `BwaMem` task as `BwaMem.``analysisReadyBam` and `BwaMem.analysisReadySorted`, respectively. Therefore, we call our next task `MarkDuplicates` with input `BwaMem.analysisReadySorted`:
```

Comment:
Therefore, we call our next task `MarkDuplicates` with input `BwaMem.analysisReadySorted` change to 
 Therefore, we call our next task `MarkDuplicates` with input `BwaMem.analysisReadySorted` from the BwaMem task